### PR TITLE
Anchor Destination Expand on Click Implementation

### PR DESCRIPTION
**Pull request by keithwrightbos**
Registers document click listener that allows for expansion for anchor href to be used for eventual navigation.  See intent to implement #3665

### CODE REVIEW

File: src/service/url-replacements-impl.js [hunk 1]
```diff
@@ -337,6 +342,28 @@ export class UrlReplacements {
   }
 
   /**
+   * Registers click listener on document that will expand the href of clicked
+   * anchors that will cause navigation (preventDefault has not been called)
```
Comment (tdrl) [inline src/service/url-replacements-impl.js]: Nit: I'm having a hard time parsing this comment.  Is this:

(registers click listener that (will expand href of anchors that (cause navigation && !preventDefault)))

or is it:

(registers click listener that (will expand href of anchors) that (cause navigation && !preventDefault)))?

File: src/service/url-replacements-impl.js [hunk 2]
```diff
@@ -337,6 +342,28 @@ export class UrlReplacements {
...
+   * with additional support for CLICK_X & CLICK_Y based on clientX & clientY.
+   * If the target is within a shadowRoot, the x/y location is offset by the
+   * shadowRoot host's offset.
+   * @private
+   */
+  registerClickListener_() {
+    if (!this.win_.document) {
```
Comment (tdrl) [inline src/service/url-replacements-impl.js]: Is this.win_ preferred, or this.getWin()?

File: src/service/url-replacements-impl.js [hunk 3]
```diff
@@ -337,6 +342,28 @@ export class UrlReplacements {
...
+      dev.warn('no document?!  testing?');
```
Comment (tdrl) [inline src/service/url-replacements-impl.js]: More informative error message?

File: src/service/url-replacements-impl.js [hunk 4]
```diff
@@ -337,6 +342,28 @@ export class UrlReplacements {
...
+      return;
+    }
+    listen(this.win_.document, 'click', evt => {
+      this.getExpandAnchorHref_(evt).then((href, error) => {
```
Comment (tdrl) [inline src/service/url-replacements-impl.js]: If you're not using error, can you omit it?

File: src/service/url-replacements-impl.js [hunk 5]
```diff
@@ -564,6 +592,109 @@ export class UrlReplacements {
...
+    }
+    if (element && element.host) {
+      return {top: element.host./*OK*/offsetTop,
+        left: element.host./*OK*/offsetLeft};
+    }
+    return {top: 0, left: 0};
+  }
+
+  /**
+   * For testing purpose only.
```
Comment (tdrl) [inline src/service/url-replacements-impl.js]: Nit: I assume you mean `@visbileForTesting`?  Or do you mean that this method exists only for testing?

File: src/service/url-replacements-impl.js [hunk 6]
```diff
@@ -564,6 +592,109 @@ export class UrlReplacements {
...
+   * @param {!Element}
+   * @return {!{top: number, left: number}}
+   * @private
+   */
+  getShadowHostOffset_(element) {
+    while(element && element.nodeType != 11) {
+      element = element.parentNode; // 11 = DOCUMENT_FRAGMENT_NODE
+    }
+    if (element && element.host) {
+      return {top: element.host./*OK*/offsetTop,
```
Comment (tdrl) [inline src/service/url-replacements-impl.js]: Are you allowed to set /*OK*/ yourself?  I thought you had to have one of the AMP project folks check it first?

File: src/service/url-replacements-impl.js [hunk 7]
```diff
@@ -564,6 +592,109 @@ export class UrlReplacements {
...
+        if (evt.clientX === undefined) {
+          return '';
+        }
+        return evt.clientX - this.getShadowHostOffset_(evt.target).left;
+      }).bind(this),
+      'CLICK_Y': (function() {
+        if (evt.clientY === undefined) {
+          return '';
+        }
+        return evt.clientY - this.getShadowHostOffset_(evt.target).top;
```
Comment (tdrl) [inline src/service/url-replacements-impl.js]: You call getShadowHostOffset_() twice, and each one climbs the DOM.  Worth calling once and caching?

File: src/service/url-replacements-impl.js [hunk 8]
```diff
@@ -564,6 +592,109 @@ export class UrlReplacements {
...
+      ampDefaultPrevented = true;
+    }
+    return timer.promise().then(() => {
+      if (ampDefaultPrevented) {
+        return Promise.resolve(null);
+      }
+      // It is possible that later click listener modified href so use
+      // current value.
+      const currHref = evt.target.getAttribute('href');
+      if (currHref != href) {
```
Comment (tdrl) [inline src/service/url-replacements-impl.js]: JS newbie paranoia: Should this be `!==`?

Comment (taymonbeal) [inline src/service/url-replacements-impl.js]: `Element.prototype.getAttribute` returns `(string|undefined)`, and `==` and `===` have identical semantics when both operands are of that type. I recall some Google style resource (not the main style guide) saying to prefer `==` over `===`, but I don't remember where I saw that and might have imagined it.

**Comment by keithwrightbos** [comment]
After communication with Malte have decided to go with the simpler approach of not allowing promise based keys to be expanded.  Will rework and open a new PR when ready.

**Pull request closed without merging**